.cbmc-batch/stubs/aws_hash_table_no_slots_override.c

```diff
@@ -143,3 +143,67 @@ int aws_hash_table_find(const struct aws_hash_table *map, const void *key, struc
     AWS_POSTCONDITION(aws_hash_table_is_valid(map), "Output hash_table [map] must be valid.");
     return AWS_OP_SUCCESS;
 }
+
+#ifdef HASH_ITER_ELEMENT_GENERATOR
+void HASH_ITER_ELEMENT_GENERATOR(struct aws_hash_iter *new_iter, const struct aws_hash_iter *old_iter);
+#endif
+
+/* Simple map for what the iterator does: it just chugs along, returns a nondet value, until its gone at most
+ * map->entry_count times */
+struct aws_hash_iter aws_hash_iter_begin(const struct aws_hash_table *map) {
+    /* Leave it as non-det as possible */
+    AWS_PRECONDITION(aws_hash_table_is_valid(map));
+
+    /* Build a nondet iterator, set the required fields, and return it */
+    struct aws_hash_iter rval;
+    rval.map = map;
+    rval.limit = map->p_impl->entry_count;
+    rval.slot = 0;
+    rval.status = (rval.slot == rval.limit) ? AWS_HASH_ITER_STATUS_DONE : AWS_HASH_ITER_STATUS_READY_FOR_USE;
+#ifdef HASH_ITER_ELEMENT_GENERATOR
+    HASH_ITER_ELEMENT_GENERATOR(&rval, NULL);
+#endif
+    return rval;
+}
+
+bool aws_hash_iter_done(const struct aws_hash_iter *iter) {
+    AWS_PRECONDITION(
+        iter->status == AWS_HASH_ITER_STATUS_DONE || iter->status == AWS_HASH_ITER_STATUS_READY_FOR_USE,
+        "Input aws_hash_iter [iter] status should either be done or ready to use.");
+    bool rval = iter->slot == iter->limit;
+    assert(rval == (iter->status == AWS_HASH_ITER_STATUS_DONE));
+    return rval;
+}
+
+void aws_hash_iter_next(struct aws_hash_iter *iter) {
+    if (iter->slot == iter->limit) {
+        iter->status = AWS_HASH_ITER_STATUS_DONE;
+        return;
+    }
+
+    /* Build a nondet iterator, set the required fields, and copy it over */
+    struct aws_hash_iter rval;
+    rval.map = iter->map;
+    rval.limit = iter->limit;
+    rval.slot = iter->slot + 1;
+    rval.status = (rval.slot == rval.limit) ? AWS_HASH_ITER_STATUS_DONE : AWS_HASH_ITER_STATUS_READY_FOR_USE;
+#ifdef HASH_ITER_ELEMENT_GENERATOR
+    HASH_ITER_ELEMENT_GENERATOR(&rval, iter);
+#endif
+
+    *iter = rval;
+}
+
+/* delete always leaves the element unusable, so we'll just leave the element totally nondet */
+void aws_hash_iter_delete(struct aws_hash_iter *iter, bool destroy_contents) {
+    /* reduce the size of the underlying map */
+    iter->map->p_impl->entry_count--;
```

Comment:
How are you making sure there is no underflow?

Reply:
Added additional checks